README.md
# screenshot_maker

A package for generating images laid out using Flutter. This package allows you to use Flutter's amazing layout system to generate images. You can use it to create screenshots for the App Store and Play Store.


![out2のコピー](https://user-images.githubusercontent.com/12369062/97675378-8e6e9c00-1ad2-11eb-8504-fcbb378326c1.png)

You can use Flutter Inspector and Debug Paint to debug the layout.

![out2 2](https://user-images.githubusercontent.com/12369062/97772099-09918a00-1b87-11eb-922e-b9be8735be5b.png)

This package provides two widgets: ScreenshotMaker and Simulated.


## Usage

### ScreenshotMaker

Renders the child according to the specified size and writes it to the specified file.

Passing this widget to runApp and launching it as a Flutter app with a simulator will output the image.
The simulator you use is independent of the output size.
In order to easily adjust the layout, I recommend using a tablet device with a large display such as an iPad.

```dart
import 'dart:io';
import 'package:flutter/material.dart';
import 'package:screenshot_maker/screenshot_maker.dart';

void main() {
  runApp(
    ScreenshotMaker(
      // TODO: write your absolute path.
      outputFile: File('/Users/your_name/Desktop/simple.png'),
      size: Size(500, 1000),
      child: MaterialApp(
        home: Scaffold(
          appBar: AppBar(),
          body: Text('text'),
        ),
      ),
    ),
  );
}
```

### Simulated

Fits your Flutter app into the device frame image you specify and compose it.
Position and size are specified by innerScreenOffset and innerScreenSize.

![show_inner_screen_offset 2](https://user-images.githubusercontent.com/12369062/97772442-8a9e5080-1b8a-11eb-8e38-b1963c350b33.png)

The device frame image must have a transparent display area.
There are no device frame images included in this package, so you'll need to prepare them.
There is a very simple one in example/assets, but I don't recommend using it in production.

```dart
import 'package:flutter/material.dart';
import 'package:screenshot_maker/screenshot_maker.dart';

void main() {
  runApp(
    Directionality(
      textDirection: TextDirection.ltr,
      child: Simulated(
        innerScreenSize: Size(1658, 3588),
        innerScreenOffset: Size(116, 103),
        originalScreenSize: Size(1242, 2688),
        deviceFrameImage: Image.asset('assets/example_device_frame.png'),
        child: MyAwesomeApp(),
      ),
    ),
  );
}

class MyAwesomeApp extends StatelessWidget {
  const MyAwesomeApp();
  @override
  Widget build(BuildContext context) {
    return MaterialApp(home: Scaffold(appBar: AppBar(), body: Text('foobar')));
  }
}
```

CHANGELOG.md
## [0.1.0] - 2021-01-28

* When Simulated is used with null arguments, it now uses the sample device frame image.

## [0.0.2] - 2020-11-01

* Added a link to the repository.
* Wrote longer description in pubspec.yaml.


## [0.0.1] - 2020-10-31

* Initial release.


test/screenshot_maker_test.dart
import 'dart:async';
import 'dart:io';

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:mockito/mockito.dart';

import 'package:screenshot_maker/screenshot_maker.dart';

void main() {
  // This test doesn't work!! Please give me some hints to test this package.
  testWidgets('check red 1 dot output', (WidgetTester tester) async {
    final completer = Completer<void>();
    final outputFile = MockFile();
    await tester.pumpWidget(
        ScreenshotMaker(
          completer: completer,
          outputFile: outputFile,
          size: Size(1, 1),
          child: MaterialApp(
            home: Container(
              width: 1,
              height: 1,
              color: Color.fromARGB(255, 255, 0, 0),
            ),
          ),
        ),
        Duration(seconds: 3));
    await tester.pump(Duration(seconds: 3));
    await expect(completer.future, completes);
    await verify(outputFile.writeAsBytesSync(any));
    // Skips this test because it doesn't work.
  }, skip: true);
}

class MockFile extends Mock implements File {}


example/lib/main.dart
import 'dart:async';
import 'dart:io';

import 'package:flutter/material.dart';

import 'package:screenshot_maker/screenshot_maker.dart';

void main() {
  runApp(MakeScreenshotsContinuously());
}

class MakeScreenshotsContinuously extends StatefulWidget {
  final sizes = <Size>[
    Size(1242, 2688),
    Size(1242, 2208),
    Size(2048, 2732),
  ];
  final outputFiles = <File>[
    // TODO: write your absolute paths.
    File('/Users/your_name/Desktop/out1.png'),
    File('/Users/your_name/Desktop/out2.png'),
    File('/Users/your_name/Desktop/out3.png'),
  ];
  MakeScreenshotsContinuously({
    Key key,
  }) : super(key: key);

  @override
  _MakeScreenshotsContinuouslyState createState() =>
      _MakeScreenshotsContinuouslyState();
}

class _MakeScreenshotsContinuouslyState
    extends State<MakeScreenshotsContinuously> {
  int index;
  Completer<void> completer;
  File outputFile;
  Size size;

  @override
  void initState() {
    super.initState();
    index = 0;
    completer = Completer();
    outputFile = widget.outputFiles[index];
    size = widget.sizes[index];
  }

  @override
  Widget build(BuildContext context) {
    completer.future.then((_) {
      if (index < widget.outputFiles.length - 1 && completer.isCompleted) {
        setState(() {
          ++index;
          completer = Completer();
          outputFile = widget.outputFiles[index];
          size = widget.sizes[index];
        });
      }
    });

    return ScreenshotMaker(
      outputFile: outputFile,
      size: size,
      completer: completer,
      child: Directionality(
        textDirection: TextDirection.ltr,
        child: Stack(
          children: [
            // background
            Container(
              width: double.infinity,
              height: double.infinity,
              decoration: BoxDecoration(
                gradient: LinearGradient(
                  begin: Alignment.topCenter,
                  end: Alignment.bottomCenter,
                  colors: [Colors.purple[200], Colors.cyan[200]],
                ),
              ),
            ),
            Column(
              children: [
                Expanded(
                  flex: 1,
                  child: Center(
                    child: Text(
                      'You Can Do Everything with This App!',
                      style: TextStyle(
                          fontSize: 60,
                          color: Colors.white,
                          decoration: TextDecoration.none),
                    ),
                  ),
                ),
                Expanded(
                  flex: 4,
                  child: Simulated(
                    innerScreenSize: Size(1658, 3588),
                    innerScreenOffset: Size(116, 103),
                    originalScreenSize: Size(1242, 2688),
                    deviceFrameImage:
                        Image.asset('assets/example_device_frame.png'),
                    child: MyAwesomeApp(),
                  ),
                ),
              ],
            ),
          ],
        ),
      ),
    );
  }
}

class MyAwesomeApp extends StatelessWidget {
  const MyAwesomeApp();
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
        debugShowCheckedModeBanner: false,
        home: Scaffold(
          appBar: AppBar(
            title: Text('My Awesome App'),
          ),
          body: Center(
              child: Text(
            'foobar',
            style: TextStyle(fontSize: 50),
          )),
        ));
  }
}


lib/screenshot_maker.dart
/// A package for generating images laid out using Flutter.
library screenshot_maker;

import 'dart:async';
import 'dart:io';
import 'dart:ui';

import 'package:flutter/rendering.dart';
import 'package:flutter/widgets.dart';

/// A widget for generating images.
/// When you pass a widget as a child and let it work in the simulator, it generates an image.
///
/// ```dart
/// import 'dart:io';
/// import 'package:flutter/material.dart';
/// import 'package:screenshot_maker/screenshot_maker.dart';
///
/// void main() {
///   runApp(
///     ScreenshotMaker(
///       // write your absolute path.
///       outputFile: File('/Users/your_name/Desktop/simple.png'),
///       size: Size(500, 1000),
///       child: MaterialApp(
///         home: Scaffold(
///           appBar: AppBar(),
///           body: Text('text'),
///         ),
///       ),
///     ),
///   );
/// }
/// ```
class ScreenshotMaker extends StatelessWidget {
  const ScreenshotMaker({
    @required this.size,
    @required this.outputFile,
    @required this.child,
    this.completer,
    Key key,
  }) : super(key: key);

  /// Size of the image to be generated.
  final Size size;

  /// The widget to be drawn as an image.
  final Widget child;

  /// The file to be output.
  final File outputFile;

  /// A completer that is completed when the image generation is finished.
  /// By responding to this, the next image generation process can be executed.
  final Completer<void> completer;

  /// A GlobalKey to access the RenderRepaintBoundary to retrieve the image.
  static GlobalKey renderRepaintBoundaryKey = GlobalKey();

  @override
  Widget build(BuildContext context) {
    WidgetsBinding.instance.addPostFrameCallback((timeStamp) async {
      // Because it takes some time to rasterize, here is an artificial delay of 2 seconds.
      // See the following link.
      // https://api.flutter.dev/flutter/flutter_driver/FlutterDriver/screenshot.html
      await Future.delayed(Duration(seconds: 2));
      if (completer == null || (completer != null && !completer.isCompleted)) {
        final rrb = ScreenshotMaker.renderRepaintBoundaryKey.currentContext
            .findRenderObject() as RenderRepaintBoundary;
        final imgData = await rrb.toImage();
        final byteData = await imgData.toByteData(format: ImageByteFormat.png);

        final pngBytes = byteData.buffer.asUint8List();
        outputFile.writeAsBytesSync(pngBytes);
        completer?.complete();
      }
    });

    // Thanks to this FittedBox, you can use Flutter Inspector and Debug Paint on a simulator.
    return FittedBox(
      child: ConstrainedBox(
        constraints: BoxConstraints(
          maxWidth: size.width,
          minWidth: size.width,
          maxHeight: size.height,
          minHeight: size.height,
        ),
        child: MediaQuery(
          data: MediaQueryData(size: size),
          child: RepaintBoundary(
            key: ScreenshotMaker.renderRepaintBoundaryKey,
            child: child,
          ),
        ),
      ),
    );
  }
}

/// A widget for fitting your app into a device frame image.
/// If all arguments except child are set to null, the sample device frame image will be used.
///
/// ```dart
/// import 'package:flutter/material.dart';
/// import 'package:screenshot_maker/screenshot_maker.dart';
///
/// void main() {
///   runApp(
///     Directionality(
///       textDirection: TextDirection.ltr,
///       child: Simulated(
///         innerScreenSize: Size(1658, 3588),
///         innerScreenOffset: Size(116, 103),
///         originalScreenSize: Size(1242, 2688),
///         deviceFrameImage: Image.asset('assets/example_device_frame.png'),
///         child: MyAwesomeApp(),
///       ),
///     ),
///   );
/// }
///
/// class MyAwesomeApp extends StatelessWidget {
///   const MyAwesomeApp();
///   @override
///   Widget build(BuildContext context) {
///     return MaterialApp(home: Scaffold(appBar: AppBar(), body: Text('foobar')));
///   }
/// }
/// ```
class Simulated extends StatelessWidget {
  const Simulated({
    @required Widget deviceFrameImage,
    @required Size innerScreenSize,
    @required Size innerScreenOffset,
    @required this.child,
    Size originalScreenSize,
    Key key,
  })  : deviceFrameImage = deviceFrameImage ??
            const Image(
                image: AssetImage('assets/sample_device_frame.png',
                    package: 'screenshot_maker')),
        innerScreenSize = innerScreenSize ?? const Size(1658, 3588),
        innerScreenOffset = innerScreenOffset ?? const Size(116, 103),
        originalScreenSize =
            originalScreenSize ?? (innerScreenSize ?? const Size(1658, 3588)),
        super(key: key);

  /// An Image widget to display the image of the device you want to combine.
  final Image deviceFrameImage;

  /// The size of the screen in the device image.
  final Size innerScreenSize;

  /// The offset from the top left corner to the screen in the device image.
  final Size innerScreenOffset;

  /// The number of original physical pixels of the device being used as the device image.
  /// If this is null, it will be the same as innerScreenSize.
  final Size originalScreenSize;

  /// Your app for combining with the device image.
  final Widget child;

  @override
  Widget build(BuildContext context) {
    final devicePixelRatio = WidgetsBinding.instance.window.devicePixelRatio;
    return FittedBox(
      child: Stack(children: [
        Padding(
          padding: EdgeInsets.only(
            left: innerScreenOffset.width,
            top: innerScreenOffset.height,
          ),
          child: SizedBox(
            width: innerScreenSize.width,
            height: innerScreenSize.height,
            child: FittedBox(
              alignment: Alignment.topLeft,
              child: ConstrainedBox(
                constraints: BoxConstraints(
                  minWidth: originalScreenSize.width / devicePixelRatio,
                  maxWidth: originalScreenSize.width / devicePixelRatio,
                  minHeight: originalScreenSize.height / devicePixelRatio,
                  maxHeight: originalScreenSize.height / devicePixelRatio,
                ),
                child: child,
              ),
            ),
          ),
        ),
        deviceFrameImage,
      ]),
    );
  }
}
